File: lib/apps/chat/chat_app.dart
import 'package:flutter/material.dart';
import '../../core/sub_app.dart';
import '../../core/search_result.dart';
import '../../services/share_content.dart';
import 'models/chat.dart';
import 'models/message.dart';
import 'screens/chat_list_screen.dart';
import 'screens/chat_detail_screen.dart';
import 'services/chat_storage.dart';

class ChatApp extends SubApp {
  @override
  String get id => 'chat';

  @override
  String get name => 'Chat';

  @override
  IconData get icon => Icons.chat;

  @override
  Color get themeColor => Colors.blue;

  @override
  Future<void> onInit() async {
    // Storage is singleton, no init needed
  }

  @override
  Future<void> onDispose() async {
    await ChatStorage.instance.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return const ChatListScreen();
  }

  @override
  List<ShareContentType> get acceptedShareTypes => [ShareContentType.text];

  @override
  Future<void> onReceiveShare(ShareContent content) async {
    if (content.type == ShareContentType.text) {
      final text = content.data['text'] as String? ?? '';
      final fileName = content.data['fileName'] as String?;
      final isFileContent = content.data['isFileContent'] as bool? ?? false;
      final isDerivativeContent = content.data['isDerivativeContent'] as bool? ?? false;

      if (text.isNotEmpty || isFileContent || isDerivativeContent) {
        // Create a new chat with the shared text as the first message
        final now = DateTime.now();
        final chatId = now.millisecondsSinceEpoch.toString();

        // Generate title based on source
        String title;
        if (isFileContent || isDerivativeContent) {
          title = fileName ?? 'Shared Content';
        } else {
          title = 'Shared: ${text.length > 30 ? '${text.substring(0, 30)}...' : text}';
        }

        final chat = Chat(
          id: chatId,
          title: title,
          createdAt: now,
          updatedAt: now,
          isTitleGenerating: false, // Don't generate title for file shares
        );
        await ChatStorage.instance.createChat(chat);

        // Create user message with the shared content
        final message = Message(
          id: '${chatId}_0',
          chatId: chatId,
          content: text.isNotEmpty ? text : 'Shared file: $fileName',
          isUser: true,
          createdAt: now,
          metadata: {
            'isFileContent': isFileContent,
            'isDerivativeContent': isDerivativeContent,
            'fileName': fileName,
          },
        );
        await ChatStorage.instance.createMessage(message);
      }
    }
  }

  @override
  bool get supportsSearch => true;

  @override
  Future<List<SearchResult>> search(String query) async {
    final results = <SearchResult>[];

    // Search chats by title
    final chats = await ChatStorage.instance.searchChats(query);
    for (final chat in chats) {
      results.add(SearchResult(
        id: chat.id,
        type: SearchResultType.chat,
        appId: id,
        title: chat.title,
        subtitle: null,
        preview: null,
        navigationData: {'chatId': chat.id},
        timestamp: chat.updatedAt,
      ));
    }

    // Search messages by content
    final messages = await ChatStorage.instance.searchMessages(query);
    for (final message in messages) {
      // Get the chat title for this message
      final chat = await ChatStorage.instance.getChat(message.chatId);

      // Create a preview from the message content
      String preview = message.content.replaceAll('\n', ' ').trim();
      if (preview.length > 100) {
        preview = '${preview.substring(0, 100)}...';
      }

      results.add(SearchResult(
        id: message.id,
        type: SearchResultType.chatMessage,
        appId: id,
        title: chat?.title ?? 'Chat',
        subtitle: message.isUser ? 'You' : 'Assistant',
        preview: preview,
        navigationData: {'chatId': message.chatId, 'messageId': message.id},
        timestamp: message.createdAt,
      ));
    }

    return results;
  }

  @override
  void navigateToSearchResult(BuildContext context, SearchResult result) async {
    final chatId = result.navigationData['chatId'] as String;
    final chat = await ChatStorage.instance.getChat(chatId);
    if (chat == null) {
      if (context.mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('Chat not found')),
        );
      }
      return;
    }
    if (context.mounted) {
      Navigator.of(context).push(
        MaterialPageRoute(
          builder: (context) => ChatDetailScreen(
            chat: chat,
            onChatUpdated: () {},
          ),
        ),
      );
    }
  }
}

File: playground/Playground/Playground/Sources/Apps/Chat/ChatApp.swift
import SwiftUI
import Combine

/// Chat sub-app - LLM-powered conversations
class ChatApp: SubApp {
    let id = "chat"
    let name = "Chat"
    let iconName = "bubble.left.and.bubble.right.fill"
    let themeColor = Color.blue

    let supportsSearch = true

    // ObservableObject conformance
    let objectWillChange = ObservableObjectPublisher()

    init() {}

    func buildView() -> AnyView {
        AnyView(ChatListView())
    }

    // MARK: - Lifecycle

    func onInit() async {
        print("✅ Chat app initialized")
    }

    // MARK: - Search

    func search(query: String) async -> [SearchResult] {
        // Search in chat titles
        let chatsResult = ChatStorage.shared.searchChats(query: query, limit: 25)
        guard let chats = chatsResult.value else {
            print("❌ Chat search failed: \(chatsResult.error?.localizedDescription ?? "unknown error")")
            return []
        }

        let chatResults = chats.map { chat in
            SearchResult(
                id: chat.id,
                type: .chat,
                appId: id,
                title: chat.title,
                subtitle: nil,
                preview: nil,
                navigationData: ["chatId": chat.id],
                timestamp: chat.updatedAt
            )
        }

        // Search in message content
        let messagesResult = ChatStorage.shared.searchMessages(query: query, limit: 25)
        guard let messages = messagesResult.value else {
            print("❌ Message search failed: \(messagesResult.error?.localizedDescription ?? "unknown error")")
            return chatResults
        }

        let messageResults = messages.compactMap { message -> SearchResult? in
            let chatResult = ChatStorage.shared.getChat(id: message.chatId)
            guard let chat = chatResult.value, let unwrappedChat = chat else {
                return nil
            }

            let preview = message.content.count > 100
                ? String(message.content.prefix(100)) + "..."
                : message.content

            return SearchResult(
                id: message.id,
                type: .chatMessage,
                appId: id,
                title: unwrappedChat.title,
                subtitle: message.role.rawValue.capitalized,
                preview: preview,
                navigationData: ["chatId": message.chatId, "messageId": message.id],
                timestamp: message.createdAt
            )
        }

        return (chatResults + messageResults)
            .sorted { ($0.timestamp ?? Date.distantPast) > ($1.timestamp ?? Date.distantPast) }
    }

    func navigateToSearchResult(result: SearchResult) async {
        // Navigation will be handled by the UI layer
        // This is a placeholder for now
    }
}
